### PR TITLE
Remove goto exit in messaging layer

### PR DESCRIPTION
**Pull request by kghost**
#### Problem
goto exit pattern should be replaced with RAII in C++ #7721

#### Change overview
Remove `goto exit` in messaging layer `src/messaging`

#### Testing
* Manually tested using unit tests

### CODE REVIEW

**Review by jmartinez-silabs — APPROVED** [review]

File: src/messaging/ExchangeHandle.cpp [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+{
+    if (mContext != nullptr)
+        mContext->Release();
+}
+
+ExchangeHandle & ExchangeHandle::operator=(const ExchangeHandle & that)
+{
+    if (this == &that)
+        return *this;
+    mContext = that.mContext;
```
Comment (andy31415) [inline src/messaging/ExchangeHandle.cpp:47]: Are we supposed to relese current mContext before assigning?

File: src/messaging/ExchangeHandle.h [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+
+    bool HasValue() const { return mContext != nullptr; }
+    void Release();
+
+    ExchangeContext * operator->() const { return mContext; }
+    bool operator==(const ExchangeHandle & that) const { return mContext == that.mContext; }
+    bool operator!=(const ExchangeHandle & that) const { return !(*this == that); }
+
+private:
+    friend class ExchangeContext;
```
Comment (andy31415) [inline src/messaging/ExchangeHandle.h:56]: Could we avoid friend somehow? maybe creating a static "fromRawExchangeContext" to be super explicit and have a `ExchangeContext *InternalGetRawContext()` if needed ?

The friend coupling of  things is not ideal.

**Review by kpschoedel — APPROVED** [review]

File: src/messaging/ReliableMessageContext.cpp [hunk 3]
```diff
@@ -181,11 +182,14 @@ CHIP_ERROR ReliableMessageContext::HandleRcvdAck(uint32_t AckMsgId)
...
 
 {
-    CHIP_ERROR err = CHIP_NO_ERROR;
-
     // Skip processing ack if drop ack debug is enabled.
     if (ShouldDropAckDebug())
-        return err;
+        return CHIP_NO_ERROR;
+
+    auto schedule = MakeDefer([&]() {
```
Comment (mspang) [inline src/messaging/ReliableMessageContext.cpp:189]: It might be simpler to just factor out the top?

```
CHIP_ERROR ReliableMessageContext::HandleNeedsAck(uint32_t MessageId, BitFlags<MessageFlagValues> MsgFlags) {
    // Skip processing ack if drop ack debug is enabled.
    if (ShouldDropAckDebug())
        return CHIP_NO_ERROR;

    CHIP_ERROR status = HandleNeedsAckInner(messageId, msgFlags);

    // Schedule next physical wakeup on function exit
    GetReliableMessageMgr()->StartTimer();

    return status;
}
```

Comment (kghost) [reply src/messaging/ReliableMessageContext.cpp:189]: I think it depends on personal preference.

From code size aspect, `defer` is inlined, so it should be smaller that an outer function.
From readability aspect, I'd like have a single function instead of 2 functions doing one thing.

The better way to write it in C++ should be using `finally` following a `try` block, but we have disabled exception, so this is not an option.

`defer` is a widely used concept introduced by golang to solve RAII problems. IMHO, it is fine to use it here.

Comment (andy31415) [reply src/messaging/ReliableMessageContext.cpp:189]: There seems to be a small code size hit here - I am guessing templated defer will duplicate code.
I would be ok to occasionally need  deferring, however for most other needs I would prefer explicit classes (e.g. some places have a logic of 'log on error' which is widely used). the "makeDefer" should be a last resort.

**Review by andy31415 — APPROVED** [review]

File: src/lib/support/Defer.h [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+public:
+    Defer(Function && function) : mFunction(std::move(function)) {}
+    ~Defer() { mFunction(); }
+
+private:
+    Function mFunction;
+};
+
+template <typename Function>
+Defer<Function> MakeDefer(Function && function)
```
Comment (andy31415) [inline src/lib/support/Defer.h:31]: This would benefit from some documentation and usage examples.

File: src/lib/support/tests/TestDefer.cpp [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+#include <support/Defer.h>
+#include <support/UnitTestRegistration.h>
+
+#include <memory>
+
+#include <nlunit-test.h>
+
+using namespace chip;
+
+static void TestDeferUsage(nlTestSuite * inSuite, void * inContext)
```
Comment (mspang) [inline src/lib/support/tests/TestDefer.cpp:28]: Please use an unnamed namespace rather than static..

File: src/messaging/ReliableMessageContext.cpp (hunk 3 above)
Comment (mspang) [reply src/messaging/ReliableMessageContext.cpp:189]: > I think it depends on personal preference.
> 
> From code size aspect, `defer` is inlined, so it should be smaller than using an outer function.

A similar argument applies to outer function.. if it's inlined then the standalone inner function is unused and will not be linked. Probably the impact is the same.

> From readability aspect, I'd like have a single function instead of 2 functions doing one thing.
> 
> The better way to write it in C++ should be using `finally` following a `try` block, but we have disabled exception, so this is not an option.
> 
> `defer` is a widely used concept introduced by golang to solve RAII problems. IMHO, it is fine to use it here.

**Review by mspang — APPROVED** [review]

File: src/lib/support/Defer.h [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+};
+
+/**
+ * @brief
+ *   Defers the execution of a function until the surrounding function returns.
+ *   Create a Defer object holding a lambda:
+ *
+ *     auto deferred = MakeDefer([]() { -- do cleanup -- });
+ *
+ *   The lambda will be called as soon as the object goes out of the scope.
```
Comment (bzbarsky-apple) [inline src/lib/support/Defer.h:44]:
```suggestion
 *   The lambda will be called as soon as the object goes out of scope.
```

File: src/lib/support/tests/TestDefer.cpp [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+{
+    bool deferred = false;
+    {
+        auto deferredFunction = MakeDefer([&]() { deferred = true; });
+    }
```
Comment (bzbarsky-apple) [inline src/lib/support/tests/TestDefer.cpp:33]:
```suggestion
        NL_TEST_ASSERT(inSuite, !deferred);
    }
```

File: src/messaging/ExchangeContext.cpp [hunk 8]
```diff
@@ -109,35 +110,34 @@ CHIP_ERROR ExchangeContext::SendMessage(Protocols::Id protocolId, uint8_t msgTyp
...
-            SuccessOrExit(err);
+            if (err != CHIP_NO_ERROR)
+            {
+                SetResponseExpected(false);
+                return err;
+            }
         }
     }
 
+    // It is required to set `err`, which is used in the deferred `handleError`
```
Comment (bzbarsky-apple) [inline src/messaging/ExchangeContext.cpp:132]: There is no handleError I see around here....

File: src/messaging/ExchangeContext.cpp [hunk 9]
```diff
@@ -109,35 +110,34 @@ CHIP_ERROR ExchangeContext::SendMessage(Protocols::Id protocolId, uint8_t msgTyp
     if (sendFlags.Has(SendMessageFlags::kExpectResponse))
     {
         // Only one 'response expected' message can be outstanding at a time.
-        VerifyOrExit(!IsResponseExpected(), err = CHIP_ERROR_INCORRECT_STATE);
+        if (IsResponseExpected())
+        {
+            return CHIP_ERROR_INCORRECT_STATE;
```
Comment (bzbarsky-apple) [inline src/messaging/ExchangeContext.cpp:115]: This seems to be fixing a bug where if we had a "response expected" message pending and tried to send another one we would fail to send the new one _and_ stop waiting for the old one.

That's a good thing to fix, but could you please add a test for this situation?  Followup is OK, as long as we check that the test fails without this PR and passes with it.

Also, it might be better to fix it as a separate PR, to keep this refactoring PR behavior-preserving....

File: src/messaging/ExchangeContext.cpp [hunk 10]
```diff
@@ -377,11 +377,10 @@ CHIP_ERROR ExchangeContext::HandleMessage(const PacketHeader & packetHeader, con
...
-    Retain();
+    ExchangeHandle ref(this);
 
-    CHIP_ERROR err = mDispatch->OnMessageReceived(payloadHeader, packetHeader.GetMessageId(), peerAddress, msgFlags,
-                                                  GetReliableMessageContext());
-    SuccessOrExit(err);
+    ReturnErrorOnFailure(mDispatch->OnMessageReceived(payloadHeader, packetHeader.GetMessageId(), peerAddress, msgFlags,
+                                                      GetReliableMessageContext()));
 
     if (IsAckPending() && !mDelegate)
```
Comment (bzbarsky-apple) [inline src/messaging/ExchangeContext.cpp:385]: The mis-merge against this block is what's causing the compilation failures....

File: src/messaging/ExchangeHandle.h [hunk 11]
```diff
@@ -0,0 +1,62 @@
...
+public:
+    explicit ExchangeHandle() : mContext(nullptr) {}
+    explicit ExchangeHandle(ExchangeContext * context);
+    ~ExchangeHandle();
+
+    ExchangeHandle(const ExchangeHandle & that);
+    ExchangeHandle(ExchangeHandle && that) : mContext(that.mContext) { that.mContext = nullptr; }
+
+    ExchangeHandle & operator=(const ExchangeHandle & that);
+    ExchangeHandle & operator=(ExchangeHandle && that)
```
Comment (bzbarsky-apple) [inline src/messaging/ExchangeHandle.h:40]: Why is this function inlined, unlike all the other ones?

File: src/messaging/ExchangeMgr.cpp [hunk 12]
```diff
@@ -266,7 +265,8 @@ void ExchangeManager::OnMessageReceived(const PacketHeader & packetHeader, const
     // an ack to the peer.
     else if (!payloadHeader.NeedsAck())
     {
-        ExitNow(err = CHIP_ERROR_UNSOLICITED_MSG_NO_ORIGINATOR);
+        ChipLogError(ExchangeManager, "OnMessageReceived failed, err = %s", ErrorStr(CHIP_ERROR_UNSOLICITED_MSG_NO_ORIGINATOR));
```
Comment (bzbarsky-apple) [inline src/messaging/ExchangeMgr.cpp:268]: If we're doing this log in this specific spot, how about we log what's really happening?  Something like:
```suggestion
        ChipLogError(ExchangeManager, "Dropping unsolicited message that claims to be a response");
```

File: src/messaging/ExchangeMgr.cpp [hunk 13]
```diff
@@ -290,7 +290,11 @@ void ExchangeManager::OnMessageReceived(const PacketHeader & packetHeader, const
             ec = mContextPool.CreateObject(this, payloadHeader.GetExchangeID(), session, false, matchingUMH->Delegate);
         }
 
-        VerifyOrExit(ec != nullptr, err = CHIP_ERROR_NO_MEMORY);
+        if (ec == nullptr)
+        {
+            ChipLogError(ExchangeManager, "OnMessageReceived failed, err = %s", ErrorStr(CHIP_ERROR_NO_MEMORY));
```
Comment (bzbarsky-apple) [inline src/messaging/ExchangeMgr.cpp:295]:
```suggestion
            ChipLogError(ExchangeManager, "Failed to allocate exchange context to handle received message");
```

File: src/messaging/ReliableMessageMgr.cpp [hunk 14]
```diff
@@ -350,34 +350,33 @@ bool ReliableMessageMgr::CheckAndRemRetransTable(ReliableMessageContext * rc, ui
...
-
-    VerifyOrReturnError(rc != nullptr, err);
-
-    // Now that we know this is a valid entry, grab the message id from the
-    // retained buffer.  We need to do that now, because we're about to hand it
-    // over to someone else, and on failure it will no longer be available.
-    msgId = entry->retainedBuf.GetMsgId();
+    if (rc == nullptr)
+    {
+        return CHIP_ERROR_INCORRECT_STATE;
```
Comment (bzbarsky-apple) [inline src/messaging/ReliableMessageMgr.cpp:356]: This is changing behavior from returning `CHIP_NO_ERROR` to returning `CHIP_ERROR_INCORRECT_STATE`.  Is that on purpose?  It might be better to do the behavior change as a separate PR...

File: src/messaging/ReliableMessageMgr.cpp [hunk 15]
```diff
@@ -350,34 +350,33 @@ bool ReliableMessageMgr::CheckAndRemRetransTable(ReliableMessageContext * rc, ui
...
 
     const ExchangeMessageDispatch * dispatcher = rc->GetExchangeContext()->GetMessageDispatch();
-    VerifyOrExit(dispatcher != nullptr, err = CHIP_ERROR_INCORRECT_STATE);
-
-    err = dispatcher->SendPreparedMessage(rc->GetExchangeContext()->GetSecureSession(), entry->retainedBuf);
-    SuccessOrExit(err);
+    if (dispatcher == nullptr)
+    {
+        ChipLogError(ExchangeManager, "Crit-err %s. when sending CHIP MsgId:%08" PRIX32 ", send tries: %d",
+                     ErrorStr(CHIP_ERROR_INCORRECT_STATE), entry->retainedBuf.GetMsgId(), entry->sendCount);
```
Comment (bzbarsky-apple) [inline src/messaging/ReliableMessageMgr.cpp:363]: The error message here could be a lot more useful than just "incorrect state", now that we know _exactly_ why we are erroring out.

**Review by bzbarsky-apple — CHANGES_REQUESTED** [review]

File: src/messaging/ExchangeHandle.h (hunk 11 above)
Comment (kghost) [reply src/messaging/ExchangeHandle.h:40]: This one doesn's depend on the implementation of `ExchangeContext`, but others depend on it.

As here `ExchangeContext` is a forward declaration, others can be not implemented here.

File: src/messaging/ExchangeMgr.cpp (hunk 12 above)
Comment (kghost) [reply src/messaging/ExchangeMgr.cpp:268]: I'm re-using the same error message to optimize code size, as the error string is typically 50-100 bytes, this is not a small impact.

File: src/messaging/ExchangeMgr.cpp (hunk 13 above)
Comment (kghost) [reply src/messaging/ExchangeMgr.cpp:295]: Same as above

File: src/messaging/ReliableMessageMgr.cpp (hunk 15 above)
Comment (kghost) [reply src/messaging/ReliableMessageMgr.cpp:363]: same as above, potential to reuse the error message.

File: src/messaging/ExchangeHandle.h (hunk 11 above)
Comment (bzbarsky-apple) [reply src/messaging/ExchangeHandle.h:40]: Right, but why do we want to inline this one at all?

File: src/messaging/ExchangeMgr.cpp (hunk 12 above)
Comment (bzbarsky-apple) [reply src/messaging/ExchangeMgr.cpp:268]: OK.  Worth a comment in the code then, so someone doesn't "improve" it.

File: src/messaging/ExchangeMgr.cpp (hunk 13 above)
Comment (bzbarsky-apple) [reply src/messaging/ExchangeMgr.cpp:295]: Again, should be documented that there's something subtle here.

**Comment by bzbarsky-apple** [comment]
@kghost This needs to be merged on top of #8050, sorry...  We'll want a cleanup function in `ExchangeContext::HandleMessage`

**Comment by kghost** [comment]
@bzbarsky-apple please take a look

File: src/messaging/ExchangeContext.cpp (hunk 9 above)
Comment (kghost) [reply src/messaging/ExchangeContext.cpp:118]: Added a comment for it

File: src/messaging/ExchangeHandle.h (hunk 11 above)
Comment (kghost) [reply src/messaging/ExchangeHandle.h]: I have added a template class `ReferenceCountedHandle` for this part. instead of `ExchangeHandle`

File: src/messaging/ExchangeMgr.cpp (hunk 12 above)
Comment (kghost) [reply src/messaging/ExchangeMgr.cpp:268]: Added a comment for it

File: src/messaging/ExchangeContext.cpp [hunk 16]
```diff
@@ -112,36 +112,37 @@ CHIP_ERROR ExchangeContext::SendMessage(Protocols::Id protocolId, uint8_t msgTyp
...
+            }
         }
     }
 
-    err = mDispatch->SendMessage(mSecureSession, mExchangeId, IsInitiator(), GetReliableMessageContext(),
-                                 reliableTransmissionRequested, protocolId, msgType, std::move(msgBuf));
-
-exit:
-    if (err != CHIP_NO_ERROR && IsResponseExpected())
     {
```
Comment (bzbarsky-apple) [inline src/messaging/ExchangeContext.cpp:135]: Might be worth a comment about why the extra scope is there; presumably to avoid shadowing warning for `err`?

File: src/messaging/ExchangeContext.cpp [hunk 17]
```diff
@@ -392,29 +393,52 @@ CHIP_ERROR ExchangeContext::HandleMessage(const PacketHeader & packetHeader, con
     bool isStandaloneAck = payloadHeader.HasMessageType(Protocols::SecureChannel::MsgType::StandaloneAck);
     bool isDuplicate     = msgFlags.Has(MessageFlagValues::kDuplicateMessage);
 
-    CHIP_ERROR err = mDispatch->OnMessageReceived(payloadHeader, packetHeader.GetMessageId(), peerAddress, msgFlags,
-                                                  GetReliableMessageContext());
-    SuccessOrExit(err);
+    auto deffered = MakeDefer([&]() {
```
Comment (bzbarsky-apple) [inline src/messaging/ExchangeContext.cpp:396]:
```suggestion
    auto deferred = MakeDefer([&]() {
```

File: src/messaging/ExchangeContext.cpp [hunk 18]
```diff
@@ -392,29 +393,52 @@ CHIP_ERROR ExchangeContext::HandleMessage(const PacketHeader & packetHeader, con
...
+        // Also don't close if there's an outer HandleMessage invocation.  It'll deal with the closing.
```
Comment (bzbarsky-apple) [inline src/messaging/ExchangeContext.cpp:397]: The comment here should be the "The alreadyHandlingMessage check is effectively a workaround" bit that's further below; otherwise this comment makes no sense and neither does the other one; this looks like a merge problem when the `alreadyHandlingMessage` bit got moved around but the comments did not follow it?

File: src/messaging/ExchangeContext.cpp [hunk 19]
```diff
@@ -392,29 +393,52 @@ CHIP_ERROR ExchangeContext::HandleMessage(const PacketHeader & packetHeader, con
...
+        if (alreadyHandlingMessage)
+            return;
```
Comment (bzbarsky-apple) [inline src/messaging/ExchangeContext.cpp:399]:
```suggestion
        if (alreadyHandlingMessage)
        {
            return;
        }
```

File: src/messaging/ExchangeContext.cpp [hunk 20]
```diff
@@ -392,29 +393,52 @@ CHIP_ERROR ExchangeContext::HandleMessage(const PacketHeader & packetHeader, con
...
+        mFlags.Clear(Flags::kFlagHandlingMessage);
+        // We are the outermost HandleMessage invocation.  We're not handling a message anymore.
```
Comment (bzbarsky-apple) [inline src/messaging/ExchangeContext.cpp:401]:
```suggestion
        // We are the outermost HandleMessage invocation.  We're not handling a message anymore.
        mFlags.Clear(Flags::kFlagHandlingMessage);
```

File: src/messaging/ExchangeContext.cpp [hunk 21]
```diff
@@ -392,29 +393,52 @@ CHIP_ERROR ExchangeContext::HandleMessage(const PacketHeader & packetHeader, con
...
+        // to that is that if we have a null mDelegate then our lifetime is not
+        // application-defined, since we don't interact with the application at that
+        // point.  That can happen when we are already closed (in which case
+        // MessageHandled is a no-op) or if we were just created to send a
+        // standalone ack for this incoming message, in which case we should treat
+        // it as an app-level message for purposes of our state.
+        //
+        // The alreadyHandlingMessage check is effectively a workaround for the fact that
+        // SendMessage() is not calling MessageHandled() yet and will go away when
+        // we fix that.
```
Comment (bzbarsky-apple) [inline src/messaging/ExchangeContext.cpp:414]: This part of the comment should move to where we actually check `alreadyHandlingMessage`.

File: src/messaging/ExchangeContext.cpp [hunk 22]
```diff
@@ -392,29 +393,52 @@ CHIP_ERROR ExchangeContext::HandleMessage(const PacketHeader & packetHeader, con
...
+        if ((isStandaloneAck || isDuplicate) && mDelegate != nullptr)
+            return;
```
Comment (bzbarsky-apple) [inline src/messaging/ExchangeContext.cpp:416]:
```suggestion
        if ((isStandaloneAck || isDuplicate) && mDelegate != nullptr)
        {
            return;
        }
```

File: src/lib/support/ReferenceCountedHandle.h [hunk 23]
```diff
@@ -0,0 +1,47 @@
...
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#pragma once
+
+namespace chip {
+
+/** ReferenceCountedHandle acts like a unique_ptr to an object derived from ReferenceCounted. In contrast to unique_ptr,
```
Comment (bzbarsky-apple) [inline src/lib/support/ReferenceCountedHandle.h:22]: But ... it doesn't act as unique_ptr?  It acts more like shared_ptr, no?

File: src/lib/support/ReferenceCountedHandle.h [hunk 24]
```diff
@@ -0,0 +1,47 @@
...
+
+    ReferenceCountedHandle(const ReferenceCountedHandle & that) : mTarget(that.mTarget) { mTarget.Retain(); }
+
+    ReferenceCountedHandle & operator=(const ReferenceCountedHandle & that) = delete;
+    ReferenceCountedHandle(ReferenceCountedHandle && that)                  = delete;
+    ReferenceCountedHandle & operator=(ReferenceCountedHandle && that) = delete;
+
+    Target & Get() const { return mTarget; }
+    Target * operator->() const { return &mTarget; }
+    bool operator==(const ReferenceCountedHandle & that) const { return mTarget == that.mTarget; }
```
Comment (bzbarsky-apple) [inline src/lib/support/ReferenceCountedHandle.h:40]: This is a somewhat odd implementation of `operator==`.  It tests object equality, not object identity, which is not what I would normally expect operator== on a handle like this to do...

Is this handle just meant to be used on the stack to keep things alive?  If so, I would propose getting rid of the Get() and operator-> and comparison operators, and perhaps deleting the copy constructor and just leaving the constructor and destructor.  And renaming to AutoHoldRef, maybe.  Or AutoRefcount.  Or something like that...  In any case, the "Handle" in the name seems superfluous: the important part is not the handle-ness but the automatic reference counting we get.

We might even want to do that if we do plan to use it in other contexts, and add features as we actually need them and can actually make sure they work correctly because we are using them.

Comment (kghost) [reply src/lib/support/ReferenceCountedHandle.h:40]: This is a mistake when I changed `mTarget` from pointer to reference, I'll fix it.

I agree that copy constructor is not necessary, and I'd prefer an explicit copy function instead of the copy constructor.

But `Get()` and `operator->` are essential for use-cases of  holding an object in another structure. Anyway I'll remove them here, and add them back when needed.

File: src/messaging/ExchangeContext.cpp [hunk 25]
```diff
@@ -112,36 +112,38 @@ CHIP_ERROR ExchangeContext::SendMessage(Protocols::Id protocolId, uint8_t msgTyp
...
-        SetResponseExpected(false);
-    }
-
-    // Release the reference to the exchange context acquired above. Under normal circumstances
-    // this will merely decrement the reference count, without actually freeing the exchange context.
-    // However if one of the function calls in this method resulted in a callback to the protocol,
-    // the protocol may have released its reference, resulting in the exchange context actually
-    // being freed here.
-    Release();
+        // Create a new scope for `err`, to avoid shadowing warning previous `err`.
```
Comment (bzbarsky-apple) [inline src/messaging/ExchangeContext.cpp:136]:
```suggestion
        // Create a new scope for `err`, to avoid warning for previous `err` shadowing this one.
```

**Review by bzbarsky-apple — APPROVED** [review]

**Comment by bzbarsky-apple** [comment]
But note that I am still very unconvinced by the `ReferenceCountedHandle` name...

**Comment by kghost** [comment]
> But note that I am still very unconvinced by the `ReferenceCountedHandle` name...

The name should be hidden behind a type define, I'd like to choose a better name, but none comes into my mind.

File: src/messaging/ReliableMessageContext.cpp (hunk 3 above)
Comment (mspang) [reply src/messaging/ReliableMessageContext.cpp:189]: Because it allows writing arbitrary code out of execute order in a surprising way.

Why do I think the defer pattern is worse?

File: src/lib/support/tests/TestDefer.cpp (hunk 5 above)
Comment (mspang) [reply src/lib/support/tests/TestDefer.cpp:28]: @kghost

**Comment by bzbarsky-apple** [comment]
> I'd like to choose a better name, but none comes into my mind.

`AutoReferenceCount`?
